code-experiments/src/coco_utilities.c

```diff
@@ -156,57 +157,72 @@ static int coco_file_exists(const char *path) {
 }
 
 /**
- * @brief Calls the right mkdir() method (depending on the platform).
+ * @brief Calls the right mkdir() method (depending on the platform) with full privileges for the user. 
+ * If the created directory has not existed before, returns 0, otherwise returns 1. If the directory has 
+ * not been created, a coco_error is raised. 
  *
  * @param path The directory path.
  *
- * @return 0 on successful completion, and -1 on error.
+ * @return 0 if the created directory has not existed before and 1 otherwise.
  */
 static int coco_mkdir(const char *path) {
+  int result = 0;
+
 #if _MSC_VER
-  return _mkdir(path);
+  result = _mkdir(path);
 #elif defined(__MINGW32__) || defined(__MINGW64__)
-  return mkdir(path);
+  result = mkdir(path);
 #else
-  return mkdir(path, S_IRWXU);
+  result = mkdir(path, S_IRWXU);
 #endif
+
+  if (result == 0)
+    return 0;
+  else if (errno == EEXIST)
+    return 1;
+  else 
+    coco_error("coco_mkdir(): unable to create %s, mkdir error: %s", path, strerror(errno));
+    return 1; /* Never reached */
 }
 
 /**
- * @brief Creates a directory with full privileges for the user.
- *
- * @note Should work cross-platform.
+ * @brief Creates a directory (possibly having to create nested directories). If the last created directory 
+ * has not existed before, returns 0, otherwise returns 1.
  *
  * @param path The directory path.
+ *
+ * @return 0 if the created directory has not existed before and 1 otherwise.
  */
-static void coco_create_directory(const char *path) {
-  char *tmp = NULL;
-  char *p;
-  size_t len = strlen(path);
+static int coco_create_directory(const char *path) {
+  char *path_copy = NULL;
+  char *tmp, *p;
   char path_sep = coco_path_separator[0];
+  size_t len = strlen(path);
 
-  /* Nothing to do if the path exists. */
-  if (coco_directory_exists(path))
-    return;
+  int result = 0;
+
+  path_copy = coco_strdup(path);
+  tmp = path_copy;
 
-  tmp = coco_strdup(path);
-  /* Remove possible trailing slash */
+  /* Remove possible leading and trailing (back)slash */
   if (tmp[len - 1] == path_sep)
     tmp[len - 1] = 0;
-  for (p = tmp + 1; *p; p++) {
+  if (tmp[0] == path_sep)
+    tmp++;
+
```

Comment:
Am I correct that this will not allow to use absolute paths under *nix systems? What if I want to write to `/home/dimo/`? In which use case is it important to remove the leading slash?

Reply:
Unless I'm mistaken, this was always meant to work only for relative paths. It's true, though, that this should be explicitly written, and it's not now.